qa: Test shared validation interface

promag opened the pull request:
Based on #18338, only last commit matters. Quoting the test description:

```
// Test UnregisterSharedValidationInterface ensuring that if interface is
// unregistered during the middle of a callback, interface is destroyed as soon
// as callback returns.
```

All credits to ryanofsky.

promag commented:
Initially this was in  #18338, decided to push a new PR because some travis job hanged.

ryanofsky commented:
I think #18524 (now merged) might fix the xenial test hang: https://travis-ci.org/github/bitcoin/bitcoin/jobs/668847707

MarcoFalke commented:
Open-Close to re-run ci. See https://github.com/bitcoin/bitcoin/issues/15847#issuecomment-487591130

sipa committed:
Do not clear validationinterface entries being executed

The previous code for MainSignalsInstance::Clear would decrement the reference
count of every interface, including ones that were already Unregister()ed but
still being executed.

promag committed:
Add RegisterSharedValidationInterface test

promag commented:
Now based on #18551, applied the following change to the test to account for @sipa fix (this now fails in master):
```diff
@@ -397,6 +397,7 @@ BOOST_AUTO_TEST_CASE(release_shared)
     BOOST_CHECK(!test_interface);
     BOOST_CHECK_EQUAL(state, CALLED);
     state = UNREGISTERED;
+    UnregisterAllValidationInterfaces();
     unregistered.set_value(true);
     BOOST_CHECK(destroyed.get_future().get());
     BOOST_CHECK_EQUAL(state, DESTROYED);
```

promag commented:
appveyor failed with "OSError: [WinError 10048] Only one usage of each socket address (protocol/network address/port) is normally permitted", restarted.

promag commented:
Please set milestone 0.20

MarcoFalke commented:
Needs rebase

fanquake commented:
@promag are you still interested in working on / making this change?
